app/src/org/commcare/android/tasks/FormRecordCleanupTask.java

```diff
@@ -77,184 +72,150 @@ public FormRecordCleanupTask(Context context, CommCarePlatform platform, int tas
         }
-        
-        System.out.println("Synced: " + unindexedRecords.size() + ". Removed: " + oldrecords + " old records, and " + (recordsToRemove.size() - oldrecords) + " busted new ones");
+
+        System.out.println("Synced: " + unindexedRecords.size() +
+                ". Removed: " + oldrecords + " old records, and " +
+                (recordsToRemove.size() - oldrecords) + " busted new ones");
+
         return SUCCESS;
     }
 
-    private int cleanupRecord(FormRecord r, SqlStorage<FormRecord> storage) {
-        try {
-            FormRecord updated = getUpdatedRecord(context, platform, r, FormRecord.STATUS_SAVED);
-            if(updated == null) {
-                return DELETE;
-            } else {
-                storage.write(updated);
+    /**
+     * Reparse the saved form instance associated with the form record and
+     * apply any updates found to the form record, such as UUID and date
+     * modified, returning an updated copy.  Write the updated record to
+     * storage. If the record is associated with a case id, recompute and write
+     * the SessionStateDescriptor too.
+     *
+     * @param context   Used to get the filepath of the form instance
+     *                  associated with the record.
+     * @param oldRecord Reparse this record and return an updated copy of it
+     * @param newStatus The new form record status
+     * @param storage   User storage where updated FormRecord is written
+     * @return The reparsed form record and the associated case id, if present
+     * @throws IOException                       Problem opening the saved form
+     *                                           attached to the record.
+     * @throws InvalidStructureException         Occurs during reparsing of the
+     *                                           form attached to record.
+     * @throws XmlPullParserException
+     * @throws UnfullfilledRequirementsException Parsing encountered a platform
+     *                                           versioning problem
+     */
+    public static FormRecord updateAndWriteRecord(Context context,
+                                                  CommCarePlatform platform,
+                                                  FormRecord oldRecord,
+                                                  String newStatus,
+                                                  SqlStorage<FormRecord> storage)
+            throws InvalidStructureException, IOException,
+            XmlPullParserException, UnfullfilledRequirementsException {
+
+        Pair<FormRecord, String> recordUpdates =
+            reparseRecord(context, oldRecord, FormRecord.STATUS_SAVED);
+        FormRecord updated = recordUpdates.first;
+        String caseId = recordUpdates.second;
+
+        if (caseId != null &&
+                FormRecord.STATUS_UNINDEXED.equals(oldRecord.getStatus()) &&
+                !FormRecord.STATUS_UNINDEXED.equals(newStatus)) {

```

Comment:
Actually realized newStatus is either SAVED or the same as oldRecord.getStatus, so in a subsequent refactor I take it out completely.